### PR TITLE
Cleanup volume information from db when deleted

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR aims at fixing the issue caused when:
- download a Volume (such that there is a record in the volume_store_ref table)
- migrate the same volume from one primary store to the another
When migrating between primary storage pools ACS temporarily copies the volume to the secondary store and then copies it from there to the destination primary store and then the secondary store volume directory is cleared up and the volume (volumes table) is marked as Expunged. BUT the corresponding volume's secondary store record(created when we downloaded the volume) still appears to be ready
This PR fixes this issue by clearing all records on the secondary store corresponding to that volume and also enhances the garbage collection logic to delete any volumes that were marked as Expunged but have entries in the volume_store_ref table.

Observations made:
Entries in volume_store_ref after downloading the volume and initiating migration:
```
*************************** 2. row ***************************
                  id: 113
            store_id: 1
           volume_id: 127
             zone_id: 1
             created: 2020-12-15 05:45:40
        last_updated: NULL
              job_id: NULL
        download_pct: 100
                size: 0
       physical_size: 0
      download_state: DOWNLOADED
            checksum: NULL
           error_str: NULL
          local_path: NULL
        install_path: volumes/2/127/a73cb20a-9ab3-4c4d-9605-31a3850e1fdc.qcow2
                 url: NULL
        download_url: http://10.1.34.20/userdata/38406061-14aa-499a-92c7-6f25978759ae.qcow2
               state: Ready
           destroyed: 0
        update_count: 2
             ref_cnt: 0
             updated: 2020-12-15 05:45:52
download_url_created: 2020-12-15 05:45:52
*************************** 3. row ***************************
                  id: 114
            store_id: 1
           volume_id: 127
             zone_id: 0
             created: 2020-12-15 05:46:10
        last_updated: NULL
              job_id: NULL
        download_pct: 0
                size: 0
       physical_size: 0
      download_state: NULL
            checksum: NULL
           error_str: NULL
          local_path: NULL
        install_path: volumes/2/127/7b5ba7e3-bc06-4c31-8eda-74cda3fab182.qcow2
                 url: NULL
        download_url: NULL
               state: Copying
           destroyed: 0
        update_count: 3
             ref_cnt: 0
             updated: 2020-12-15 05:46:20
download_url_created: NULL
```

Once migration completes, the directory is cleared up from the secondary store, the record corresponding to the migrating volume is also cleared up - but the entry for the downloaded volume is still left behind:
```
*************************** 2. row ***************************
                  id: 113
            store_id: 1
           volume_id: 127
             zone_id: 1
             created: 2020-12-15 05:45:40
        last_updated: NULL
              job_id: NULL
        download_pct: 100
                size: 0
       physical_size: 0
      download_state: DOWNLOADED
            checksum: NULL
           error_str: NULL
          local_path: NULL
        install_path: volumes/2/127/a73cb20a-9ab3-4c4d-9605-31a3850e1fdc.qcow2
                 url: NULL
        download_url: http://10.1.34.20/userdata/38406061-14aa-499a-92c7-6f25978759ae.qcow2
               state: Ready
           destroyed: 0
        update_count: 2
             ref_cnt: 0
             updated: 2020-12-15 05:45:52
download_url_created: 2020-12-15 05:45:52
2 rows in set (0.00 sec)


MariaDB [cloud]> select id,account_id,name,folder,path,volume_type,created,updated,removed,state from volumes where id = 127\G
*************************** 1. row ***************************
         id: 127
 account_id: 2
       name: ROOT-46
     folder: /acs/primary/ref-trl-1977-k-M7-pearl-dsilva/ref-trl-1977-k-M7-pearl-dsilva-kvm-pri1
       path: 87ff45f2-e466-42f2-9f75-b772fc26a4f1
volume_type: ROOT
    created: 2020-12-15 05:33:09
    updated: 2020-12-15 05:46:30
    removed: NULL
      state: Expunged
1 row in set (0.00 sec)

```
### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### How Has This Been Tested?
The same steps mentioned above were repeated with the fix in place and the following are observed:
Firstly, the garbage collector identifies any left behind volume entries and clears them
```
020-12-15 06:07:14,285 DEBUG [c.c.s.StorageManagerImpl] (StorageManager-Scavenger-1:ctx-c21cc4c6) (logid:95f250bd) Secondary storage garbage collector found 1 volumes to cleanup on volume_store_ref for store: NFS://10.10.0.16/acs/secondary/ref-trl-1977-k-M7-pearl-dsilva/ref-trl-1977-k-M7-pearl-dsilva-sec1
2020-12-15 06:07:14,286 DEBUG [c.c.s.StorageManagerImpl] (StorageManager-Scavenger-1:ctx-c21cc4c6) (logid:95f250bd) Deleting volume store DB entry: VolumeDataStore[113-127-1volumes/2/127/a73cb20a-9ab3-4c4d-9605-31a3850e1fdc.qcow2]
```

Downloaded a volume:
```
MariaDB [cloud]> select * from volume_store_ref\G
​                  id: 134
            store_id: 2
           volume_id: 121
             zone_id: 1
             created: 2020-12-17 05:48:52
        last_updated: NULL
              job_id: NULL
        download_pct: 100
                size: 0
       physical_size: 0
      download_state: DOWNLOADED
            checksum: NULL
           error_str: NULL
          local_path: NULL
        install_path: volumes/2/121/afaf7b46-59d4-4e1b-b592-d1ce16d60b14.qcow2
                 url: NULL
        download_url: http://10.1.34.15/userdata/8f111dea-e1f7-4351-9155-6ec43487019a.qcow2
               state: Ready
           destroyed: 0
        update_count: 2
             ref_cnt: 0
             updated: 2020-12-17 05:49:04
download_url_created: 2020-12-17 05:49:04
2 rows in set (0.00 sec)

MariaDB [cloud]> select id,account_id,name,folder,path,volume_type,created,updated,removed,state from volumes where id = 121\G
*************************** 1. row ***************************
         id: 121
 account_id: 2
       name: ROOT-51
     folder: /acs/primary/ref-trl-1977-k-M7-pearl-dsilva/ref-trl-1977-k-M7-pearl-dsilva-kvm-pri2
       path: 2e30ed4b-9bdd-4e97-8f45-8051a1e28b16
volume_type: ROOT
    created: 2020-12-14 11:54:15
    updated: 2020-12-17 05:49:03
    removed: NULL
      state: Ready
1 row in set (0.00 sec)
```
Initiate migration - here we see another entry created for the same volume:
```
*************************** 2. row ***************************
                  id: 134
            store_id: 2
           volume_id: 121
             zone_id: 1
             created: 2020-12-17 05:48:52
        last_updated: NULL
              job_id: NULL
        download_pct: 100
                size: 0
       physical_size: 0
      download_state: DOWNLOADED
            checksum: NULL
           error_str: NULL
          local_path: NULL
        install_path: volumes/2/121/afaf7b46-59d4-4e1b-b592-d1ce16d60b14.qcow2
                 url: NULL
        download_url: http://10.1.34.15/userdata/8f111dea-e1f7-4351-9155-6ec43487019a.qcow2
               state: Ready
           destroyed: 0
        update_count: 2
             ref_cnt: 0
             updated: 2020-12-17 05:49:04
download_url_created: 2020-12-17 05:49:04
*************************** 3. row ***************************
                  id: 135
            store_id: 2
           volume_id: 121
             zone_id: 0
             created: 2020-12-17 05:50:10
        last_updated: NULL
              job_id: NULL
        download_pct: 0
                size: 0
       physical_size: 0
      download_state: NULL
            checksum: NULL
           error_str: NULL
          local_path: NULL
        install_path: volumes/2/121
                 url: NULL
        download_url: NULL
               state: Creating
           destroyed: 0
        update_count: 1
             ref_cnt: 0
             updated: 2020-12-17 05:50:10
download_url_created: NULL
```

Post migration - notice there is no entry left behind for the expunged volume:
```
MariaDB [cloud]> select * from volume_store_ref where volume_id = 121 \G
Empty set (0.00 sec)
```

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by weizhouapache** [comment]
@Pearl1594 
as far as I remember, there is a thread in StorageManager to cleanup the download urls and remove the Expunged volumes.
I suggest (1) not to change that behavior, (2) add a global setting to indicate whether remove a volume from secondary storage when remove it.

**Comment by Pearl1594** [comment]
@weizhouapache thanks for the pointers
1. The logic for cleaning up the volume on expiration of the download url is still as is, but this just removes any dangling volume entries left behind immediately after migration of volumes takes place just in case there exists any entries for that volume on the secondary store 
2. Since we are only clearing up what has been marked as destroyed / expunged, would it be necessary to enclose it within a global setting?

File: engine/storage/src/main/java/org/apache/cloudstack/storage/image/db/VolumeDataStoreDaoImpl.java [hunk 1]
```diff
@@ -120,6 +120,11 @@ public boolean configure(String name, Map<String, Object> params) throws Configu
         uploadVolumeStateSearch.and("destroyed", uploadVolumeStateSearch.entity().getDestroyed(), SearchCriteria.Op.EQ);
         uploadVolumeStateSearch.done();
 
+        allVolumesSearch = this.createSearchBuilder();
```
Comment (weizhouapache) [inline engine/storage/src/main/java/org/apache/cloudstack/storage/image/db/VolumeDataStoreDaoImpl.java:123]: it looks same as storeVolumeSearch

**Comment by weizhouapache** [comment]
> @weizhouapache thanks for the pointers
> 
> 1. The logic for cleaning up the volume on expiration of the download url is still as is, but this just removes any dangling volume entries left behind immediately after migration of volumes takes place just in case there exists any entries for that volume on the secondary store
> 2. Since we are only clearing up what has been marked as destroyed / expunged, would it be necessary to enclose it within a global setting?

@Pearl1594 thanks.
sorry I did not test the storage migration. could you tell me what to do on old volume when it is copied to another pool ? Destroy or expunge it ?

**Comment by Pearl1594** [comment]
@weizhouapache Post migration the volume is pretty much non-existent hence is marked as expunged and is removed from the secondary store entirely. But the only problem is it doesn't clean up the volume_store_ref table in case it was previously downloaded (and url hasn't expired) or there are some entries due to some failed migration, etc. So now we just ensure that the db is up to date

**Comment by weizhouapache** [comment]
> @weizhouapache Post migration the volume is pretty much non-existent hence is marked as expunged and is removed from the secondary store entirely. But the only problem is it doesn't clean up the volume_store_ref table in case it was previously downloaded (and url hasn't expired) or there are some entries due to some failed migration, etc. So now we just ensure that the db is up to date

@Pearl1594 thanks !
in storage migration , volume will be moved between primary pools. it should not impact the snapshots/volumes on secondary storage I think.

I have quick reviewed the process in AncientDataMotionStrategy.java, it copies the volume by the steps
(1) copy volume to image store
```
                answer = copyObject(srcData, objOnImageStore);
```
(2) copy from image store to another pool
```
                CopyCommand cmd = new CopyCommand(objOnImageStore.getTO(), addFullCloneFlagOnVMwareDest(destData.getTO()), _copyvolumewait, VirtualMachineManager.ExecuteInSequence.value());
```
(3) remove volume from image store
```
            imageStore.delete(objOnImageStore);
```

step (3) will remove not only the volume created in step(1), but also the volume snapshots created when download the volume in the past. I think it is a bug.

**Comment by Pearl1594** [comment]
yes @weizhouapache - that's right, but even if we prevent deletion of the volume created prior to migration, wouldn't it be unusable as we've lost reference to the old volume because we've marked that volume as expunged?

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@Pearl1594 is this ready for review/test?

@blueorangutan test

**Comment by Pearl1594** [comment]
@shwstppr This issue seems to be only present for Xen and KVM, and not on VMWare, so it needs some testing against VMware.

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java [hunk 2]
```diff
@@ -392,15 +405,16 @@ protected Answer copyVolumeBetweenPools(DataObject srcData, DataObject destData)
                     return answer;
                 }
             } catch (Exception e) {
-                if (imageStore.exists(objOnImageStore)) {
+                if (objOnImageStore != null) {
```
Comment (DaanHoogland) [inline engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:408]: is not null enough? what if it is an object but doesn't exist? or it does but not on `imageStore`?

**Review by DaanHoogland — COMMENTED** [review]
can you extract the try-catch block you touched in AcientDataMotionStrategy and give it a useful name?
one remarks otherwise.

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java (hunk 2 above)
Comment (Pearl1594) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:408]: imagestore.exists() eventually calls volumeDataStoreDao.findByStoreVolume(...) for volumes, which in case of multiple entries could possibly return the wrong record

Comment (DaanHoogland) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:408]: ok, but still, the `objOnImageStore` may be a valid object that represents nothing on our `imageStore`, can it? shouldn't we protect against that as well?

**Comment by rhtyd** [comment]
Ping @DaanHoogland @shwstppr @davidjumani @Spaceman1984 @sureshanaparti @harikrishna-patnala - help with review/testing?

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java (hunk 2 above)
Comment (Pearl1594) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:408]: @DaanHoogland I didn't quite understand your previous comment. Could you please explain what you mean by it. Thanks

Comment (harikrishna-patnala) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:408]: @Pearl1594 what you are saying might be true but during catch() we have to check if DB entry really exists or not and process the image storage object by marking it OperationFailed. objOnImageStore will always be != null (_"DataObject objOnImageStore = imageStore.create(srcData);"_) since no where in try{} block we are setting to null or changing the object.

@DaanHoogland in that method we are dealing only with database entries. If you are asking if it represents nothing on imagestore "physically" then the answer is we don't need to have concern about that.

**Comment by Pearl1594** [comment]
@blueorangutan package

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java [hunk 3]
```diff
@@ -392,15 +405,16 @@ protected Answer copyVolumeBetweenPools(DataObject srcData, DataObject destData)
...
                     objOnImageStore.processEvent(Event.OperationFailed);
                     imageStore.delete(objOnImageStore);
```
Comment (sureshanaparti) [inline engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:410]: if any volumes on store at this point, have to remove  them. check if you have to call _deleteVolumeOnSecondaryStore()_ in this case as well

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java (hunk 3 above)
Comment (rhtyd) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:410]: @Pearl1594 can you reply ^^

Comment (Pearl1594) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:410]: It makes sense to delete the volume on secondary store on failure. Will add that - needs testing.

**Comment by DaanHoogland** [comment]
@Pearl1594 a volume related error there ^^ can you check?

**Comment by rhtyd** [comment]
Ping @Pearl1594 anything to do here?
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
code looks good. I don't think we can test every corner case and it does at least do what it says on the tin.

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java (hunk 3 above)
Comment (sureshanaparti) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:410]: @Pearl1594 any update on the changes / testing ?

Comment (Pearl1594) [reply engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/AncientDataMotionStrategy.java:410]: Haven't had the chance to get to testing this yet @sureshanaparti  - Will do it by end of this week.

**Review by rhtyd — APPROVED** [review]
